### DISCUSSION TITLE
Question regarding AES-256 GCM and partial decryption

### DISCUSSION THREAD

**Discussion by DarioSamo** [openssl/openssl#22270] [Q&A]
I'm beginning to learn the ropes around encryption/decryption in general so apologies in advance if this is a silly question.

I've been able to successfully implement encryption and decryption of an entire file using the libcrypto API with the EVP_aes_256_gcm() cipher, but I'm unable to figure out how to decrypt specific blocks of the ciphertext that I want without having to use EVP_DecryptUpdate() for at least the amount of bytes for the offset that I want to read. I can skip parts of the ciphertext even by feeding the function garbage data, and the rest of the ciphertext decrypts just fine (although it obviously invalidates the tag verification, but that is not my priority at the moment).

I've initialized the algorithm with one of the test vectors provided in the examples for the key and the IV. I've read some conflicting information on whether it's possible to do what I want to do or not by just modifying the IV that I use in EVP_DecryptInit_ex. Any attempts at modifying the IV have failed to produce what I want so far.

A quick memory inspection into the context used by the GCM cipher inside the library shows me an incrementing counter that starts at 2, which matches most of the documentation I've read about the algorithm so far. However, I see no way to access this counter directly in the current stable OpenSSL API (1.1.1f). I've verified that increasing this counter in memory actually produces the result that I want, but that is obviously not the proper way to do it.

In short, I'm looking for a way to increase this counter (or alter the IV if that works) without having to run the cipher with garbage data so I can do partial decryption. If there's no way to do this, I might opt for doing a regular complete encryption/decryption on bigger, fixed size blocks.

Thank you for your time.

**Comment by bernd-edlinger** [comment]
never use the same key again.

**Comment by DarioSamo** [comment]
> never use the same key again.

Sorry, I don't see how that is related to my question. If you say because I'm using the one included in the example, I'm just doing that for prototyping. I don't think changing the key would solve the problem I described above tho.

**Comment by mattcaswell** [comment]
> I've been able to successfully implement encryption and decryption of an entire file using the libcrypto API with the EVP_aes_256_gcm() cipher, but I'm unable to figure out how to decrypt specific blocks of the ciphertext that I want without having to use EVP_DecryptUpdate() for at least the amount of bytes for the offset that I want to read. I can skip parts of the ciphertext even by feeding the function garbage data, and the rest of the ciphertext decrypts just fine (although it obviously invalidates the tag verification, but that is not my priority at the moment).

It seems to me you are trying to use AES-256-GCM to do something it is just not designed to do. The whole point of using an AEAD mode like GCM is that you get encryption *and* integrity in one go. It is *not* designed for random access in the way you are attempting. You are supposed to decrypt and verify the tag at the same time (and ignore any content where the tag does not verify).

So the question I have is - why are you using GCM  at all? Either integrity is important to you for your use case (in which case you just can't do what you are trying to do) - or it is not (perhaps because you have already assured integrity in some other way) in which case GCM is probably the wrong tool for the job.

**Comment by DarioSamo** [comment]
First of all, thank you for your detailed answer @mattcaswell 

> So the question I have is - why are you using GCM at all? Either integrity is important to you for your use case (in which case you just can't do what you are trying to do) - or it is not (perhaps because you have already assured integrity in some other way) in which case GCM is probably the wrong tool for the job.

I don't really have a hard reason to use GCM, it's just the first one I'm experimenting with and I've had a hard time finding out which one is most suited for random access due to conflicting information I've found online from other users. I can verify the integrity in another way if necessary in a previous step, so it's not necessary at all for the algorithm to guarantee that.

Which alternative would you recommend? Would CTR be a better fit for what I need to do then?

**Comment by bernd-edlinger** [comment]
Yes, but have you understood my advice about using the key only once?

**Comment by mattcaswell** [comment]
> Yes, but have you understood my advice about using the key only once?

I think you mean IV here.

**Comment by mattcaswell** [comment]
> Which alternative would you recommend? Would CTR be a better fit for what I need to do then?

My recommendation would be to employ a cryptographer to help you design a solution. I am a programmer (not a cryptographer) and I cannot know the full context of what you are trying to achieve. So, I won't make specific recommendations on what you should do.

However, I can say that CTR mode does support random access decryption. As does CBC.

**Comment by DarioSamo** [comment]
> My recommendation would be to employ a cryptographer to help you design a solution. I am a programmer (not a cryptographer) and I cannot know the full context of what you are trying to achieve. So, I won't make specific recommendations on what you should do.

That's fine. Like I said, I'm just trying to learn how some of the algorithms work so I can confirm which technical requirements are possible depending on the chosen scheme before designing a solution. Before I started, I actually assumed that none of the current algorithms supported random access. Having a clear confirmation that the one I was investigating is not intended for random access despite the internet saying it can is enough for me. 👍 

> Yes, but have you understood my advice about using the key only once?
> I think you mean IV here.

If you actually meant IV, I assume you meant that for the last proposal on my original question? I figured that'd be the case since everything I've read so far confirms reusing an IV is a pretty bad idea, so it'd add the overhead of needing a new IV for each new block.

**Comment by paulidale** [comment]
If you care about the _integrity_ of the data, you should **not** use either GCM or CTR mode without an integrity check.  GCM has this built in but CTR needs it too.

I've been asked about using both modes sufficiently often that I came up with _Pauli's Mutual Fund_.  

> Your money is safe.  All your data are encrypted using the latest government approved methods.

```
$ echo "Pauli is owed $    1000 dollars" | openssl enc -k fnord -e -aes-128-ctr -pbkdf2 >ciphertext
$ xxd ciphertext >data
$ cat data
00000000: 5361 6c74 6564 5f5f 0766 9237 007c 4b52  Salted__.f.7.|KR
00000010: ac81 4d00 79b2 1691 1f37 e1b8 fb26 203f  ..M.y....7...& ?
00000020: 5535 d358 be98 b049 9a93 3c02 ff51 20fb  U5.X...I..<..Q .
$ vi data
$ cat data
00000000: 5361 6c74 6564 5f5f 0766 9237 007c 4b52  Salted__.f.7.|KR
00000010: ac81 4d00 79b2 1691 1f37 e1b8 fb26 203f  ..M.y....7...& ?
00000020: 4c25 c359 be98 b049 9a93 3c02 ff51 20fb  U5.X...I..<..Q .
$ xxd -r data >ciphertext
$ openssl enc -k fnord -d -aes-128-ctr -pbkdf2
Pauli is owed $ 9000000 dollars
```

If you run the same commands, you will end up with different ciphertext.  Don't worry, flip the same six bits as I did in the first four bytes of the final row of data.  In my example these were originally `5535 d358` and I changed them to `4c25 c359`.

This example is for CTR mode but GCM is essentially identical if you omit the tag.

**Comment by mattcaswell** [comment]
Nice illustration of the dangers of omitting integrity checks!

**Comment by bernd-edlinger** [comment]
He said:

>  I can verify the integrity in another way if necessary in a previous step, so it's not necessary at all for the algorithm to guarantee that.

Therefore the was not a concern for me.
But I hope he knows how to ensure integrity.

**Comment by DarioSamo** [comment]
> But I hope he knows how to ensure integrity.

Yeah, I'm aware of the dangers that a modified ciphertext can produce different but readable results just by running the decryption alone without verifying its integrity in some way. How important this is is obviously dependent on the context of what it's trying to solve.

I just wasn't clear enough on the fact that GCM only does this verification after processing the entire message and not for each block that is decoded. I guess the fact that you only use the tag at the actual end of the decryption should've given that away, but it wasn't obvious enough to me quite yet.

**Comment by paulidale** [comment]
The _can't tell if something got changed until the very end_ is the reason the OpenSSL `enc` application doesn't support GCM mode.

**Comment by paulidale** [comment]
> I can verify the integrity in another way if necessary in a previous step, so it's not necessary at all for the algorithm to guarantee that.

Can you verify it in a cryptographically secure manner?

For example, CRCs and checksums are considered easy to bypass.

**Comment by paulidale** [comment]
> I'm aware of the dangers that a modified ciphertext can produce different but readable results

Other cipher modes don't suffer from this: CFB would randomise a modified block, CBC mode randomises the modified block but the change is predictable for the subsequent block.

Use cryptographic algorithms the way they were intended, don't try to modify or improve them :grin:

**Comment by DarioSamo** [comment]
> Can you verify it in a cryptographically secure manner?
> For example, CRCs and checksums are considered easy to bypass.

I need to discuss this with others first, but I wanted to make sure whether it was possible to keep both integrity and confidentiality in the same scheme while retaining the capability of random access. After all this discussion, it's pretty clear to me there's not much reason to use an scheme like GCM unless I need to guarantee both checks every time I read the data, which is currently not my case. Like you said, use the algorithm for what it's intended to.

**Comment by kroeckx** [comment]
I don't think there is a standardized method that provides both random access and integrity.

**Comment by paulidale** [comment]
> I don't think there is a standardized method that provides both random access and integrity.

Me either.

There are modes with random access and corruption on change: e.g. XTS (aimed for disc encryption) or CFB (fast random access read, slow sequential write).  Everything is a trade off.

**Comment by bernd-edlinger** [comment]
Yes, and the application where that might be needed, might be video streams.
You can skip a few frames, but how to ensure integrity then....

**Comment by mattcaswell** [comment]
DTLS is a possible model. Integrity is guaranteed on a record level. Data sent over DTLS is divided into records, and each record is individually integrity checked. DTLS packets can be delivered out-of-order or dropped altogether - and this does not alter the ability to integrity check each record. It does not offer integrity over the complete set of all records (e.g. a man-in-the-middle could maliciously cause certain packets to be dropped, but any subsequent packets that are delivered would still successfully pass integrity checks).

**Comment by DarioSamo** [comment]
XTS and CFB are a couple of interesting starting points to look at solutions, since my research is certainly related to encryption on persistent storage of large files. They've led me to a lot of interesting articles regarding the topic.

I guess I should've worded my question from the start asking what problem I intended to solve instead of trying to wrestle with an algorithm that is not designed for it, but I do have a better understanding of the problem now.
